File: Demo/RunLoop/RunLoop/ViewController.m
//
//  ViewController.m
//  RunLoop
//
//  Created by Hayder on 16/9/16.
//  Copyright © 2016年 wangzhenhai. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (nonatomic, strong) NSThread *thread;

@property (nonatomic, strong) dispatch_source_t timer;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
//    
//    [self timer];
    
//    UIImageView *imageview = [[UIImageView alloc] init];

//    self.thread = [[NSThread alloc] initWithTarget:self selector:@selector(run) object:nil];
//    
//    [self.thread start];
    
    //获得队列
    dispatch_queue_t queue = dispatch_get_global_queue(0, 0);
    
    //创建一个定时器(dispatch_source_t本质是一个OC对象)
    self.timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
    
    //设置定时器的各种属性(几时开始任务，每个多长时间执行一次)
    //GCD的时间参数，一般是纳秒(1秒 == 10的9次方纳秒)
    dispatch_time_t start = dispatch_time(DISPATCH_TIME_NOW, 3.0 * NSEC_PER_SEC);//何时开始第一个任务
    dispatch_time_t interval = 2.0 * NSEC_PER_SEC; //何时开始重复任务
    dispatch_source_set_timer(self.timer, start, interval, 0);
    
    //设置回调
    dispatch_source_set_event_handler(self.timer, ^{
       
        NSLog(@"-------%@",[NSThread currentThread]);
    });
    
    dispatch_resume(self.timer);
    
}

- (void)timer1
{
    //调用了scheduledTimerWithTimeInterval返回的定时器，已经被添加到runloop中了NSDefaultRunLoopMode
    NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:2.0f target:self selector:@selector(run) userInfo:nil repeats:NO];
    
    //修改模式
    [[NSRunLoop currentRunLoop] addTimer:timer forMode:NSRunLoopCommonModes];

////    //等价于
//    NSTimer *timer = [NSTimer timerWithTimeInterval:2.0 target:self selector:@selector(run) userInfo:nil repeats:YES];
//    
//    //定时器只运行在NSDefaultRunLoopMode下，一旦runloop进入其他模式，这个定时器就不会工作
//    [[NSRunLoop currentRunLoop] addTimer:timer forMode:NSDefaultRunLoopMode];
//    
//    //定时器会跑在标记为common modes的模式下
//    //标记为common modes的模式：NSDefaultRunLoopMode 和 UITrackRunLoopMode
//    [[NSRunLoop currentRunLoop] addTimer:timer forMode:NSRunLoopCommonModes];
    
   
}

- (void)observe
{
    //创建一个observe
    CFRunLoopObserverRef observe = CFRunLoopObserverCreateWithHandler(CFAllocatorGetDefault(), kCFRunLoopAllActivities, YES, 0, ^(CFRunLoopObserverRef observer, CFRunLoopActivity activity) {
        
        NSLog(@"%zd",activity);
    });
    
    //添加观察者：监听RunLoop的状态
    CFRunLoopAddObserver(CFRunLoopGetCurrent(), observe, kCFRunLoopDefaultMode);
    
    //释放observe
    CFRelease(observe);
}

- (void)run
{
    [[NSRunLoop currentRunLoop] addPort:[NSPort port] forMode:NSDefaultRunLoopMode];
    [[NSRunLoop currentRunLoop] run];
}


- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [self performSelector:@selector(test) onThread:self.thread withObject:nil waitUntilDone:NO];
}

- (void)test
{
    NSLog(@"%@----thread",[NSThread currentThread]);
    NSLog(@"test");
}
/**
 *  CF的内存管理(core Foundation)
 1.凡是带有create,copy,retain等字眼的函数，创建出来的对象，都需要在最后做一次release
  *比如CFRunLoopObserveCreate
 
 2.release函数
 CFRelease(对象);
 */


@end

File: Demo/Swift/语法/2.Swift3.0-可选项/2.Swift3.0-可选项/ViewController.swift
//
//  ViewController.swift
//  2.Swift3.0-可选项
//
//  Created by Hayder on 16/9/20.
//  Copyright © 2016年 wangzhenhai. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        Demo()
    }
    
    func Demo1() {
        
        //1>定义 y,没有初始化
        let y: Int?
        
        //2>给常量设置数值，初始化的工作能做一次
        y = 20;
        
        
        print(y)
        
        
        // *** var 的可选值默认为nil
        // *** let 的可选值没有默认值
        
        var x: Int?
        
        x = 10
        x = 100
        
        print(x)
        
    }
    //最常见的错误
    //unexpectedly found nil while unwrapping an Optional value
    //在解包的时候发现了nil值
    /**
        定义 可选项使用 ‘?’
        解包使用 ‘！’，准备计算
     */
    func Demo() {
        
        //1.原始的可选项定义
        //none 没有值，或者 some 某一类值
        let x: Optional = 10
        
        //2.简单定义
        //'?' 用来定义y是一个可选的Int类型，可能没有值，也可能是一个整数
        // 常量y 使用之前必须初始化
        let y: Int? = 20
        
        //输出结果Optional(10), 提示这是一个可选值
        print(x)
        print(y)
        
        //1> 不同类型之间的值不能直接运算! 如果没有值是 nil 不是任何数据类型，不能参与计算
        //print(x + y)
        //2> "!" 强行解包 - 从可选值中强行获取对应的非空值，如果是nil，就会奔溃
        // 程序员必须为每一个！负责
        // 程序中少用 ！
        print(x! + y!)
    }
}

